### DISCUSSION TITLE
Flexible, Open Source User Management for UDAN

### DISCUSSION THREAD

**Discussion by kcramakrishna** [Digital-Assistant/Digital_Assistant_Client#272] [Ideas]
The Digital Assistant will need user mgmt to encompass multiple scenarios:

1. Individual users of our SaaS offering (create account, login with Oauth2 etc)
2. MSMEs and Enterprises using our SaaS (SSO integration, delegated user mgmt)
3. MSMEs and Enterprises with on-prem and private clouds (Kubernetes/docker?) (SSO integration, fine grained user mgmt)

We need to have user mgmt. which will be flexible enough to handle all these use cases.

We have heard good things about keycloak - https://www.keycloak.org/
It has quite a few FOSS competitors.
https://syncope.apache.org/
https://solutionsreview.com/identity-management/the-10-best-free-and-open-source-identity-management-tools/
https://www.openiam.com/
https://medevel.com/iam-systems-10-identity/

Some proprietary comparisions:
https://cloudinfrastructureservices.co.uk/best-keycloak-alternatives-open-source-sso/
https://www.gartner.com/reviews/market/access-management/vendor/red_hat/product/keycloak/alternatives
https://www.g2.com/products/keycloak/competitors/alternatives

We should compare it to other FOSS Identity mgmt products/projects and finalize one that fits all of our above scenarios with minimal coding.
@yureshwar  has spent considerable time evaluating various alternatives. He can update his findings in this thread.
@venkatczn  has extensive experience with Liferay User mgmt which is pretty robust. He can take keycloak for a spin and compare it with Liferay's features and post his finding belows. Please feel free to add as many tables and as much text as you want. We can then decide on which one to go forward with.

@yureshwar, @venkatczn : Please feel free to update as much information as you can manage. Do not hesitate to mention even the smallest detail. We hope that your analyses will be useful to other FOSS architects in designing their products and solutions.

**Comment by kcramakrishna** [comment]
Keycloak integration with node

https://medium.com/javarevisited/keycloak-integrating-with-nodejs-and-angular-27eada9c8272

**Comment by yureshwar** [comment]
We can integrate keycloak with quarkus and the keycloak has been using quarkus as default distribution
https://www.keycloak.org/migration/migrating-to-quarkus
https://quarkus.io/guides/security-keycloak-authorization

Keycloak can handle multiple SSO scenarios and can be easily integrated into on premise docker also.

**Comment by yureshwar** [comment]
We should add the authentication to react application for keycloak as given here

https://blog.logrocket.com/implement-keycloak-authentication-react/

**Comment by kcramakrishna** [comment]
Here are some options for keylocak themese:
Free themes
https://github.com/p2-inc/keycloak-themes
https://github.com/lukin/keywind



https://github.com/topics/keycloak-theme 

Some infor on how to manage themse.
https://www.baeldung.com/spring-keycloak-custom-themes
https://dev.to/trigo/how-to-customize-keycloak-themes-545 
https://medium.com/keycloak/create-a-custom-theme-for-keycloak-8781207be604

**Reply by kcramakrishna** [reply]
https://www.keycloakify.dev/?lang=en 
https://github.com/keycloakify/keycloakify

**Comment by kcramakrishna** [comment]
Here is something for feature management i.e. enabling/disabling features based on flags. 

You can enable/disable features for organisations, roles, geographies, account type etc.
A 'role' can be viewed as a 'collection of flags'.

https://openfeature.dev/ - this seems too heavy but needs more reading up.

We should look at this tool to manage our feature complexity. 
1. This could be overkill. Needs detailed review.
2. Check if this has other open source alternatives. 
3. Eventually find one which is simple but flexible and robust and can last us a while.

**Reply by kcramakrishna** [reply]
Started looking for open source feature flagging tools. Used Bing/CoPilot to do some research. I somehow could not link it here.

https://posthog.com/blog/best-open-source-feature-flag-tools 
https://theproductmanager.com/tools/best-feature-flag-software/ 
https://www.bdrennan.com/posts/2020/07/25/Feature-Flags-What-Why-When-How/
https://dev.to/flagship/how-to-implement-feature-flags-in-a-react-app-1dk2
https://dev.to/cosmicflood/how-to-choose-the-right-open-source-feature-flags-management-tool-4aon

Here is an excerpt:
The following feature flagging providers mentioned earlier are true open-source solutions that you can self-host:

1. **PostHog**: An all-in-one platform that includes feature flags, product analytics, session replay, and more. It is open-source and can be self-hosted.
2. **Unleash**: A popular open-source feature flag service with comprehensive SDK support and a strong community. It is designed for self-hosting.
3. **GrowthBook**: Provides feature flags and A/B testing capabilities, with a focus on flexibility and ease of use. It is open-source and can be self-hosted.
4. **Flagsmith**: Offers feature flagging and remote configuration, with support for various programming languages. It is open-source and can be self-hosted.
5. **Flipt**: A simple and flexible feature flagging solution that can be self-hosted.

**Reply by kcramakrishna** [reply]
I think we should pick on which is easy to integrate into Java (Quarkus which is our backend) and/or React/Typescript which is our frontend.

**Comment by venkatczn** [comment]
This looks interesting. I will explore this.

We have a similar inhouse built module that supports feature flags for multi-tenancy and org-hierarchy, managing the flags both at the table and cache level. Also, wrapped by graphql and AWS appsync for syncing settings across multiple devices.


________________________________

**Reply by kcramakrishna** [reply]
Can you try to find if there are any simpler, self-hosted alternatives ?

**Comment by kcramakrishna** [comment]
https://research.google/pubs/zanzibar-googles-consistent-global-authorization-system/
https://authzed.com/blog/what-is-google-zanzibar
